compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp

```diff
@@ -271,24 +216,56 @@ struct SetBatchMatmulEncoding : public OpRewritePattern<linalg::BatchMatmulOp> {
         padAndSetEncoding(rewriter, loc, origOut, user,
                           IREE::LinalgExt::EncodingRole::RESULT, operandTypes);
 
-    Value matmulTiled = rewriter
-                            .create<linalg::BatchMatmulOp>(
-                                loc, encodedOut.getType(),
-                                ValueRange{encodedLhs, encodedRhs}, encodedOut)
-                            .getResult(0);
+    Value opTiled;
+    switch (user) {
+    case IREE::LinalgExt::EncodingUser::MATMUL:
+      opTiled = rewriter
+                    .create<linalg::MatmulOp>(
+                        loc, encodedOut.getType(),
+                        ValueRange{encodedLhs, encodedRhs}, encodedOut)
+                    .getResult(0);
+      break;
+
+    case IREE::LinalgExt::EncodingUser::BATCH_MATMUL:
+      opTiled = rewriter
+                    .create<linalg::BatchMatmulOp>(
+                        loc, encodedOut.getType(),
+                        ValueRange{encodedLhs, encodedRhs}, encodedOut)
+                    .getResult(0);
+      break;
+    case IREE::LinalgExt::EncodingUser::VECMAT:
+      opTiled = rewriter
+                    .create<linalg::VecmatOp>(
+                        loc, encodedOut.getType(),
+                        ValueRange{encodedLhs, encodedRhs}, encodedOut)
+                    .getResult(0);
+      break;
+    case IREE::LinalgExt::EncodingUser::MATVEC:
+      opTiled = rewriter
+                    .create<linalg::MatvecOp>(
+                        loc, encodedOut.getType(),
+                        ValueRange{encodedLhs, encodedRhs}, encodedOut)
+                    .getResult(0);
+      break;
+    case IREE::LinalgExt::EncodingUser::BATCH_MATVEC:
+      opTiled = rewriter
+                    .create<linalg::BatchMatvecOp>(
+                        loc, encodedOut.getType(),
+                        ValueRange{encodedLhs, encodedRhs}, encodedOut)
+                    .getResult(0);
+      break;
+    }
```

Comment:
We can do the trick like this:

https://github.com/tensorflow/mlir-hlo/blob/b3eeedc11c3f5a708a4847d211bf1f4fc738683e/mhlo/transforms/legalize_to_linalg/legalize_to_linalg.cc#L3825-L3841